### PR TITLE
Fix global type loading on Windows paths

### PR DESCRIPTION
**Pull request by tomv564**
This PR now contains 3 fixes for Windows path handling caused by two inconsistencies

The first inconsistency existed between `expectedFiles` (unix paths) and `typeRoots` (windows paths)

**Typings from e.g. jasmine/mocha not resolved by Typescript**

Repro: Open any test file in this project, receive diagnostics about not finding `it` and `describe`
Cause: Files were loaded in memory, but `ensureBasicFiles` passed windows paths into unix-style path filters (`isGlobalTsFile`).
Fix: Convert to unix paths before passing to filter methods in `ensureBasicFiles` (https://github.com/sourcegraph/javascript-typescript-langserver/commit/7f57e21bc71fa02b9533ebe0fafeb608a5ab7dd9)
Test: ✅ Added in https://github.com/sourcegraph/javascript-typescript-langserver/pull/362/commits/6d9d3d0f34d5354e924d25468cbd2a15376fec30

After this fix, `isExpectedDeclarationFile` now expected a unixPath, causing a new issue:

**Typings from typeRoots not resolved by Typescript**

Repro: Open a project using typeRoots, receive diagnostics about not finding types from typeRoots
Cause: Files from typeRoots were not loaded into memory by `ensureConfigDependencies` as it was still passing Windows paths into the now unix-path-based `isConfigDependency`
Fix: Make `typeRoots` unix-path based so it is consistent with checks against `expectedFiles`. (https://github.com/sourcegraph/javascript-typescript-langserver/commit/f0c8e02fdd58465e415adbd673151a45952bdf38)
Test: ✅ Updated to check host in https://github.com/sourcegraph/javascript-typescript-langserver/pull/362/commits/6d9d3d0f34d5354e924d25468cbd2a15376fec30

This fixed the type resolution issues I had been seeing while working on Windows. :tada:

To add test coverage, I made the ProjectManager tests use different rootUris. This did not cover the above issues, but it did uncover another inconsistency:

**Windows `rootPath` not handled correctly for configuration lookups in ProjectManager**

When ProjectManager is constructed with a Windows `rootPath`:
* It is not trimmed correctly (see `trimmedRootPath`)
* It puts Windows paths into `this.configs` which otherwise uses unix paths. 

Impact: My guess is default configurations did not work on windows, as they were windows-style and the rest of the code assumed unix-style paths were used..
Fix: Handle `rootPath` and configuration lookups in a platform-sensitive way (because there is no need to normalise to unix paths here?)

### CODE REVIEW

**Comment by tomv564** [comment]
@felixfbecker would you want to have a look at this? Not a clean diff, but I hope the comments above help explain.

The main design issue I tried to solve is knowing when windows paths are converted to unix paths, my guidelines for solving the inconsistency were:
* ProjectManager uses Windows paths as it interacts more with the filesystem.
* ProjectConfiguration uses unixPaths as it works closely with the typescript configuration/library which use unix paths on windows.

File: src/project-manager.ts [hunk 1]
```diff
@@ -929,19 +925,19 @@ export class ProjectManager implements Disposable {
...
         }
-        const rootPath = this.rootPath.replace(/\/+$/, '')
+        const rootPath = this.rootPath.replace(/[\\\/]+$/, '')
         while (dir && dir !== rootPath) {
             config = configs.get(dir)
             if (config) {
                 return config
             }
-            const pos = dir.lastIndexOf('/')
+            const pos = dir.includes('\\') ? dir.lastIndexOf('\\') : dir.lastIndexOf('/')
```
Comment (felixfbecker) [inline src/project-manager.ts]: I'd use `string.search()`

File: src/test/project-manager.test.ts [hunk 2]
```diff
@@ -3,134 +3,177 @@ import chaiAsPromised = require('chai-as-promised')
...
 import { InMemoryFileSystem } from '../memfs'
 import { ProjectManager } from '../project-manager'
+import { uri2path } from '../util'
 import { MapFileSystem } from './fs-helpers'
 chai.use(chaiAsPromised)
 const assert = chai.assert
 
 describe('ProjectManager', () => {
+    for (const rootUri of ['file:///', 'file:///c:/foo/bar/', 'file:///foo/bar/']) {
+        testWithRootUri(rootUri)
```
Comment (felixfbecker) [inline src/test/project-manager.test.ts]: why not inline this here?

**Comment by felixfbecker** [comment]
> * ProjectManager uses Windows paths as it interacts more with the filesystem.
> * ProjectConfiguration uses unixPaths as it works closely with the typescript configuration/library which use unix paths on windows.

Could you make this documentation available in the source code somewhere, e.g. in the docblock of the respective classes?

File: src/project-manager.ts (hunk 1 above)
Comment (tomv564) [reply src/project-manager.ts]: Definitely more compact:
```javascript
filePath.search(/[\\\/][^\\\/]*$/)
```
Do you prefer this? If so, I propose storing the regex as a static member somewhere (or adding this functionality to `util.ts`.

**Comment by tomv564** [comment]
For visibility I propose adding it on every docblock that describes a field/parameter.

I wish typing could help with this, eg
```typescript
type UnixPath: string
```
An option is to use `interface UnixPath extends string`, but this requires code changes.
What do you prefer?

File: src/test/project-manager.test.ts (hunk 2 above)
Comment (tomv564) [reply src/test/project-manager.test.ts]: Fixed (if I understood correctly)

**Comment by tomv564** [comment]
@felixfbecker your suggestions were applied, ready for your review again!

File: src/project-manager.ts [hunk 3]
```diff
@@ -24,6 +24,8 @@ import {
     uri2path
 } from './util'
 
+const LastForwardOrBackwardSlash = /[\\\/][^\\\/]*$/
```
Comment (felixfbecker) [inline src/project-manager.ts]: use CONSTANT_CASE

**Pull request merged**